Implicit conversions of boost::optional to bool causes compilation failure in ALCPeakFittingPresenter.

Compilation of the target `MantidScientificInterfacesMuon` fails with the following error. I believe this is because of the fact that as of boost 1.58.0 implicit conversion from `boost::optional` to `bool` is no longer valid (the conversion operator is marked `explicit`) see [here](http://www.boost.org/doc/libs/1_58_0/libs/optional/doc/html/boost_optional/synopsis.html). This means that `boost::optional` is [contextually convertible](http://en.cppreference.com/w/cpp/language/implicit_conversion) to `bool` rather than implicitly convertible in the more traditional sense. This, in turn means that it *cannot* be compared to an integer without an *explicit conversion to bool* which is not done in this implementation of glibc.

```/home/ejb/Code/Work/Mantid/master/repo/qt/scientific_interfaces/Muon/ALCPeakFittingPresenter.cpp:73:3: error: invalid operands to binary expression ('boost::optional<QString>' and 'int')
  assert(index);
  ^~~~~~~~~~~~~
/usr/include/assert.h:101:26: note: expanded from macro 'assert'
  ((void) sizeof ((expr) == 0), __extension__ ({                        \
                  ~~~~~~ ^  ~
/usr/include/QtCore/qglobal.h:2138:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'QBool' for 1st argument
inline bool operator==(QBool b1, bool b2) { return !b1 == !b2; }
            ^
/usr/include/QtCore/qbytearray.h:529:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QByteArray' for 1st argument
inline bool operator==(const QByteArray &a1, const char *a2)
            ^
/usr/include/QtCore/qstring.h:1238:32: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QStringRef' for 1st argument
inline QT_ASCII_CAST_WARN bool operator==(const QStringRef &s1, const char *s2)
                               ^
/usr/include/QtCore/qchar.h:392:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'QChar' for 1st argument
inline bool operator==(QChar c1, QChar c2) { return c1.unicode() == c2.unicode(); }
            ^
/usr/include/QtCore/qbytearray.h:527:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QByteArray' for 1st argument
inline bool operator==(const QByteArray &a1, const QByteArray &a2)
            ^
/usr/include/QtCore/qbytearray.h:531:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const char *' for 1st argument
inline bool operator==(const char *a1, const QByteArray &a2)
            ^
/usr/include/QtCore/qstring.h:909:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'QString::Null' for 1st argument
inline bool operator==(QString::Null, const QString &s) { return s.isNull(); }
            ^
/usr/include/QtCore/qstring.h:936:32: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const char *' for 1st argument
inline QT_ASCII_CAST_WARN bool operator==(const char *s1, const QString &s2)
                               ^
/usr/include/QtCore/qstring.h:1202:20: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QStringRef' for 1st argument
Q_CORE_EXPORT bool operator==(const QStringRef &s1,const QStringRef &s2);
                   ^
/usr/include/QtCore/qstring.h:1205:20: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QString' for 1st argument
Q_CORE_EXPORT bool operator==(const QString &s1,const QStringRef &s2);
                   ^
/usr/include/QtCore/qstring.h:1208:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QStringRef' for 1st argument
inline bool operator==(const QStringRef &s1,const QString &s2)
            ^
/usr/include/QtCore/qstring.h:1212:20: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QLatin1String' for 1st argument
Q_CORE_EXPORT bool operator==(const QLatin1String &s1, const QStringRef &s2);
                   ^
/usr/include/QtCore/qstring.h:1236:32: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const char *' for 1st argument
inline QT_ASCII_CAST_WARN bool operator==(const char *s1, const QStringRef &s2)
                               ^
/usr/include/boost/optional/detail/optional_relops.hpp:131:6: note: candidate function not viable: no known conversion from 'int' to 'boost::none_t' for 2nd argument
bool operator == ( optional<T> const& x, none_t ) BOOST_NOEXCEPT
     ^
/usr/include/boost/blank.hpp:58:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const boost::blank' for 1st argument
inline bool operator==(const blank&, const blank&)
            ^
/usr/include/QtCore/qglobal.h:2139:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'bool' for 1st argument
inline bool operator==(bool b1, QBool b2) { return !b1 == !b2; }
            ^
/usr/include/QtCore/qglobal.h:2140:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'QBool' for 1st argument
inline bool operator==(QBool b1, QBool b2) { return !b1 == !b2; }
            ^
/usr/include/QtCore/qstring.h:908:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'QString::Null' for 1st argument
inline bool operator==(QString::Null, QString::Null) { return true; }
            ^
/usr/include/QtCore/qstring.h:910:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QString' for 1st argument
inline bool operator==(const QString &s, QString::Null) { return s.isNull(); }
            ^
/usr/include/QtCore/qstring.h:949:32: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const char *' for 1st argument
inline QT_ASCII_CAST_WARN bool operator==(const char *s1, const QLatin1String &s2)
                               ^
/usr/include/QtCore/qstring.h:962:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QLatin1String' for 1st argument
inline bool operator==(const QLatin1String &s1, const QLatin1String &s2)
            ^
/usr/include/QtCore/qstring.h:1215:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QStringRef' for 1st argument
inline bool operator==(const QStringRef &s1,const QLatin1String &s2)
            ^
/usr/include/QtCore/qpoint.h:161:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QPoint' for 1st argument
inline bool operator==(const QPoint &p1, const QPoint &p2)
            ^
/usr/include/QtCore/qpoint.h:329:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QPointF' for 1st argument
inline bool operator==(const QPointF &p1, const QPointF &p2)
            ^
/usr/include/QtCore/qsize.h:156:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QSize' for 1st argument
inline bool operator==(const QSize &s1, const QSize &s2)
            ^
/usr/include/QtCore/qsize.h:314:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QSizeF' for 1st argument
inline bool operator==(const QSizeF &s1, const QSizeF &s2)
            ^
/usr/include/QtCore/qrect.h:496:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QRect' for 1st argument
inline bool operator==(const QRect &r1, const QRect &r2)
            ^
/usr/include/QtCore/qrect.h:833:13: note: candidate function not viable: no known conversion from 'boost::optional<QString>' to 'const QRectF' for 1st argument
inline bool operator==(const QRectF &r1, const QRectF &r2)
            ^
/usr/include/boost/optional/detail/optional_relops.hpp:62:6: note: candidate template ignored: deduced conflicting types for parameter 'T' ('QString' vs. 'int')
bool operator == ( optional<T> const& x, T const& y )
     ^
/usr/include/boost/smart_ptr/shared_ptr.hpp:790:40: note: candidate template ignored: could not match 'shared_ptr' against 'optional'
template<class T, class U> inline bool operator==(shared_ptr<T> const & a, shared_ptr<U> const & b) BOOST_NOEXCEPT
                                       ^
/usr/include/boost/smart_ptr/shared_ptr.hpp:813:31: note: candidate template ignored: could not match 'shared_ptr' against 'optional'
template<class T> inline bool operator==( shared_ptr<T> const & p, boost::detail::sp_nullptr_t ) BOOST_NOEXCEPT
                              ^
/usr/include/boost/smart_ptr/shared_ptr.hpp:818:31: note: candidate template ignored: could not match 'shared_ptr<type-parameter-0-0>' against 'int'
template<class T> inline bool operator==( boost::detail::sp_nullptr_t, shared_ptr<T> const & p ) BOOST_NOEXCEPT
                              ^
/usr/include/QtCore/qscopedpointer.h:174:13: note: candidate template ignored: could not match 'QScopedPointer' against 'optional'
inline bool operator==(const QScopedPointer<T, Cleanup> &lhs, const QScopedPointer<T, Cleanup> &rhs)
            ^
/usr/include/boost/optional/detail/optional_relops.hpp:28:6: note: candidate template ignored: could not match 'optional<type-parameter-0-0>' against 'int'
bool operator == ( optional<T> const& x, optional<T> const& y )
     ^
/usr/include/boost/optional/detail/optional_relops.hpp:96:6: note: candidate template ignored: could not match 'optional<type-parameter-0-0>' against 'int'
bool operator == ( T const& x, optional<T> const& y )
     ^
/usr/include/boost/optional/detail/optional_relops.hpp:165:6: note: candidate template ignored: could not match 'optional<type-parameter-0-0>' against 'int'
bool operator == ( none_t , optional<T> const& y ) BOOST_NOEXCEPT
     ^
1 error generated.
make[3]: *** [qt/scientific_interfaces/Muon/CMakeFiles/MantidScientificInterfacesMuon.dir/build.make:369: qt/scientific_interfaces/Muon/CMakeFiles/MantidScientificInterfacesMuon.dir/ALCPeakFittingPresenter.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/Makefile2:4828: qt/scientific_interfaces/Muon/CMakeFiles/MantidScientificInterfacesMuon.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:4840: qt/scientific_interfaces/Muon/CMakeFiles/MantidScientificInterfacesMuon.dir/rule] Error 2
make: *** [Makefile:1163: MantidScientificInterfacesMuon] Error 2
```

### Platforms affected
- Fedora 26
- clang++ version 4.0.0 and g++ version 7.1.1
- boost 1.58+
- glibc 2.25-8

--- a/qt/scientific_interfaces/Muon/ALCPeakFittingPresenter.cpp
+++ b/qt/scientific_interfaces/Muon/ALCPeakFittingPresenter.cpp
@@ -70,7 +70,7 @@ void ALCPeakFittingPresenter::onPeakPickerChanged() {
   // If PeakPicker is changed, it should be enabled, which means a peak function
   // should be selected
   // (See onCurrentFunctionChanged)
-  assert(index);
+  assert(bool(index));
 
   auto peakFunc = m_view->peakPicker();
 

